Refine OpenAI serving entrypoint to remove batch requests

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

- #7108

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

This pull request streamlines the OpenAI serving layer by eliminating batch request handling from its core entrypoints. This significant refactoring aims to simplify the codebase, making it more maintainable and easier to understand, while also incorporating specific improvements to streaming responses and tool call management.

- API Refinement: Refactored OpenAI serving entrypoints (for chat, completions, and embeddings) to exclusively handle single requests, removing previous support for batch processing.
- Code Simplification: Simplified internal request conversion logic (_convert_to_internal_request) across all serving entrypoints by removing batch-specific loops and parameters.
- Streaming Improvements: Restructured streaming response generation in chat and completion serving into dedicated asynchronous generator methods for cleaner code, and improved tool call ID generation for streaming.
- Test Updates: Updated unit tests to align with the new single-request model and adopted asyncio.run for proper execution of asynchronous tests.

<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JustinTong0323 committed:
Refactors OpenAI serving entrypoint, remove batch requests and unused validation.

This commit refactors the OpenAI serving entrypoints
to handle single requests instead of lists of requests.

This simplifies the code and makes it easier to
understand and maintain.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into refactor_openai_serving_remove_batch

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -457,55 +378,132 @@ async def _handle_streaming_request(
-                    is_first = is_firsts.get(index, True)
-                    stream_buffer = stream_buffers.get(index, "")
-                    n_prev_token = n_prev_tokens.get(index, 0)
+                prompt_tokens = content["meta_info"]["prompt_tokens"]
```
is there a reason that we removed `is_firsts`, `stream_buffers`, and `n_prev_tokens`?

CatherineSue replied:
Why do we not need the status trackers?

JustinTong0323 replied:
we have:
<img width="380" alt="image" src="https://github.com/user-attachments/assets/436a4610-1bd1-4e77-a003-405a663376ec" />

CatherineSue replied:
What if request.n > 1? index will be the index of each response.

Is a single var instead of a dictionary enough?

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -651,16 +567,14 @@ async def _handle_non_streaming_request(
         except ValueError as e:
             return self.create_error_response(str(e))
 
+        # Tokenizer manager should always return a list for chat requests
```
We can remove this comment. I checked that when `request.n==1`, ret will be a dictionary. When `request.n>1`, it will be a list.

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -894,6 +808,9 @@ async def _process_tool_call_stream(
         # Yield tool calls
         for call_item in calls:
+            # Fix tool call ID handling for streaming
+            tool_call_id = f"call_{uuid.uuid4().hex[:24]}"
```
I think this is incorrect. The calls id is assigned at the beginning of each tool call. 

For instance:

```
[{"index": 0, "id": "call_DdmO9pD3xa9XTPNJ32zg2hcA", "function": {"arguments": "", "name": "get_weather"}, "type": "function"}]
[{"index": 0, "id": null, "function": {"arguments": "{\"", "name": null}, "type": null}]
[{"index": 0, "id": null, "function": {"arguments": "location", "name": null}, "type": null}]
[{"index": 0, "id": null, "function": {"arguments": "\":\"", "name": null}, "type": null}]
[{"index": 0, "id": null, "function": {"arguments": "Paris", "name": null}, "type": null}]
[{"index": 0, "id": null, "function": {"arguments": ",", "name": null}, "type": null}]
[{"index": 0, "id": null, "function": {"arguments": " France", "name": null}, "type": null}]
[{"index": 0, "id": null, "function": {"arguments": "\"}", "name": null}, "type": null}]
null
```

This change right now assigns an id for each item?

JustinTong0323 committed:
Matches OpenAI streaming usage behavior

Ensures that streaming responses including usage data are formatted to match OpenAI's API. Specifically, when `include_usage` is true, it first sends a chunk with `finish_reason` but no usage, followed by a chunk with usage but empty choices. This aligns with OpenAI's specification for streaming responses.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_completions.py:
```diff
@@ -327,14 +245,14 @@ async def _handle_non_streaming_request(
         except ValueError as e:
             return self.create_error_response(str(e))
 
+        # Tokenizer manager should always return a list for completion requests
```
Same, remove this line

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -523,29 +523,63 @@ async def _generate_chat_stream(
-                usage=usage,
-            )
-            yield f"data: {final_chunk.model_dump_json()}\n\n"
+                # Standard behavior when not including usage
```
Looks like L563 to L582 is the same as L527 to L545?

JustinTong0323 replied:
Yes you r right, solved.

JustinTong0323 committed:
Reorders stream chunks for OpenAI API compatibility

Reorders the streaming chunks to first send a chunk with the
finish reason, followed by a separate chunk with usage information
when `include_usage` is specified. This change aligns the streaming
behavior with the OpenAI API.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 commented on python/sglang/srt/entrypoints/openai/serving_base.py:
```diff
@@ -37,7 +37,7 @@ async def handle_request(
             # Convert to internal format
             adapted_request, processed_request = self._convert_to_internal_request(
-                [request], [self._generate_request_id_base(request)]
+                request, self._generate_request_id_base(request)
```
Ok, so we just ignored it and solve later?

JustinTong0323 replied:
Removed.

JustinTong0323 committed:
Fixes tool call ID handling for streaming

Fixes an issue where tool call IDs were not handled correctly during streaming.

The tool call ID is now generated only once per tool call, and subsequent chunks use null ID and name for argument deltas. This ensures correct identification and handling of tool calls in streaming scenarios.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Removes unnecessary code branches

Removes code branches that are no longer needed
due to changes in the tokenizer manager's behavior.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Update python/sglang/srt/entrypoints/openai/serving_embedding.py

Co-authored-by: Chang Su <csu272@usc.edu>

JustinTong0323 committed:
fix

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Updates stream buffer logic

Ensures the stream_buffer is always updated with the latest state.

This change addresses a potential issue where the stream buffer might not reflect the most recent updates, particularly when dealing with delta information in streaming responses. By consistently updating the stream buffer, the code maintains a correct and consistent state throughout the streaming process.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Removes request_id from internal request conversion

Refactors the `_convert_to_internal_request` methods in the OpenAI serving classes to remove the `request_id` parameter.

The request ID is no longer needed during the conversion process, simplifying the method signatures and internal logic.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into refactor_openai_serving_remove_batch

JustinTong0323 committed:
remove rid in tests

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 commented:
local test:
<img width="1056" alt="image" src="https://github.com/user-attachments/assets/3558b1c3-ad29-4ede-a411-cc53a160dad9" />

JustinTong0323 committed:
Merge branch 'main' into refactor_openai_serving_remove_batch

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix bad merge

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Fixes streaming chat completion for multiple requests

Addresses an issue in streaming chat completion where incorrect state was maintained when handling multiple concurrent requests.

The changes ensure that state variables like `is_first`, `stream_buffer`, `n_prev_token`, `prompt_tokens`, `completion_tokens`, and `cached_tokens` are properly tracked for each request by using dictionaries indexed by the request index.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into refactor_openai_serving_remove_batch

JustinTong0323 committed:
fix: Handles multiple streams in completion endpoint

Refactors the OpenAI completion streaming endpoint to correctly
handle multiple streams. This change addresses an issue where
concurrent streams were not being managed independently, leading
to incorrect output. It introduces per-stream buffers and token
counters to ensure accurate and isolated responses for each stream.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix: Initialize content in DeltaMessage for assistant role

Updates the DeltaMessage instantiation in the OpenAI serving chat to include an empty content string for the assistant role. This change ensures that the message structure is correctly formed for processing in the chat completion response stream.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

CatherineSue commented:
> Hi, I was effectively utilizing OpenAI's batch inference feature, but is it getting deprecated?

Hi @justHungryMan , the original online batch support is not production-wise ready. We plan to switch to an offline batch support, please check out #7427 for details. Welcome to leave comments there!
